### PR TITLE
Hotfix/cdap 4790 make table apis consistent

### PR DESCRIPTION
**Pull request by anwar6953**
The makes any Table requests where an empty set of columns is specified to mean 'no columns'.
Previously, there was inconsistent behavior - this meant get all columns in some instances and meant get no columns in other instances.

https://issues.cask.co/browse/CDAP-4790
http://builds.cask.co/browse/CDAP-RBT633-11

https://wiki.cask.co/display/CE/Changes+to+Table+Interface

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 1]
```diff
@@ -127,24 +133,44 @@ public void startTx(Transaction tx) {
...
+
     try {
-      Result[] results = hTable.get(hbaseGets);
-      return Lists.transform(Arrays.asList(results), new Function<Result, Map<byte[], byte[]>>() {
-        @Nullable
-        @Override
-        public Map<byte[], byte[]> apply(Result result) {
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      List<Map<byte[], byte[]>> results =
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: Is this needed? Can't you just apply the transform in the else branch of the next loop?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 2]
```diff
@@ -127,24 +133,44 @@ public void startTx(Transaction tx) {
...
+
+      List<Map<byte[], byte[]>> toReturn = new ArrayList<>(gets.size());
+      int hbaseResultsIndex = 0;
+      for (int i = 0; i < gets.size(); i++) {
+        co.cask.cdap.api.dataset.table.Get get = gets.get(0);
+        List<byte[]> cols = get.getColumns();
+        if (cols != null && cols.isEmpty()) {
+          toReturn.add(ImmutableMap.<byte[], byte[]>of());
+        } else {
+          toReturn.add(results.get(hbaseResultsIndex++));
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: In the last comment I meant here.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 3]
```diff
@@ -127,24 +133,44 @@ public void startTx(Transaction tx) {
...
         }
-      });
+      }
+      Preconditions.checkArgument(toReturn.size() == gets.size());
+      Preconditions.checkArgument(hbaseResultsIndex == hbaseGets.size());
+      for (Map<byte[], byte[]> map : toReturn) {
+        Preconditions.checkNotNull(map);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: can this ever be null? Based on your logic above, I don't see how

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 4]
```diff
@@ -296,6 +322,7 @@ private void setFilterIfNeeded(ScanBuilder scan, @Nullable Filter filter) {
     }
   }
 
+  // null columns and empty columns both mean to retrieve all columns
   private Get createGet(byte[] row, @Nullable byte[][] columns) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: this should not allow empty cols. In the code that calls it, you only call if it is not empty. So I guess the comment above is inaccurate.

**Comment by anew** [comment]
Hi @anwar6953, I know this is not ready yet but I made some comments :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java (hunk 3 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: I wrote these preconditions as validation of the code, during development. I will remove them now, since they won't happen, based upon the code logic above.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java (hunk 4 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: updated the comment and added a Preconditions.checkArgument.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 5]
```diff
@@ -29,12 +30,16 @@
...
 
   public byte[] getRow() {
     return row;
   }
 
+  /**
+   * @return a list of columns of this object. null indicates ALL columns.
+   */
+  @Nullable
   public List<byte[]> getColumns() {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Move this method after all constructors.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 6]
```diff
@@ -43,16 +48,19 @@
 
   public RowColumns(byte[] row) {
     this.row = row;
-    this.columns = new ArrayList<>();
   }
 
   public RowColumns(byte[] row, byte[]... columns) {
     this(row);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Better not to call `this`, since the idea is to have this constructor represents some columns (potentially zero), while the other constructor is for all columns.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 7]
```diff
@@ -43,16 +48,19 @@
...
+    this.columns = new ArrayList<>();
     add(columns);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: no need to call `add`. Simply do `this.columns = Arrays.asList(columns);`.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 8]
```diff
@@ -43,16 +48,19 @@
...
   }
 
   @SuppressWarnings("unchecked")
   public T add(byte[]... columns) {
+    if (this.columns == null) {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: This actually mean if someone do this:

```java
Delete delete = new Delete(row);  // Delete all columns
delete.add();    // It is valid to be called without any arg, but this turn the delete to delete no column
```

I think it's better to guard against empty arg call by having the signature as `T add(byte[] firstColumn, byte[]... moreColumns)`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 9]
```diff
@@ -127,24 +133,44 @@ public void startTx(Transaction tx) {
...
-      Result[] results = hTable.get(hbaseGets);
-      return Lists.transform(Arrays.asList(results), new Function<Result, Map<byte[], byte[]>>() {
-        @Nullable
-        @Override
-        public Map<byte[], byte[]> apply(Result result) {
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      List<Map<byte[], byte[]>> results =
+        Lists.transform(Arrays.asList(hTable.get(hbaseGets)), new Function<Result, Map<byte[], byte[]>>() {
+          @Nullable
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: Remove, it never return `null`.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 8 above)
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: This doesn't allow passing a `new byte[][]{ C1, C2, C3 }`, though. Is that fine?

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Although doing so breaks the API (since it's possible to without arg now), although I doubt if anyone is doing that.

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: You can have a overloaded method that takes a `byte[][]` explicitly as well.

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: But it is possible that someone computes a set of rows and wants to add that. That would be awkward with the suggested API. (Our own IndexedTable does that, right?)

Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: A method that takes in `byte[][]` as Terence suggested would handle that, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java (hunk 1 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: Moved the transformation to the else branch.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 10]
```diff
@@ -29,30 +31,30 @@
...
 
   public RowColumns(byte[] row, byte[]... columns) {
-    this(row);
-    add(columns);
+    this.row = row;
+    this.columns = Arrays.asList(columns);
+  }
+
+  @SuppressWarnings("unchecked")
+  public T add(byte[] firstColumn, byte[]... moreColumns) {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: We should have Javadoc for all public methods and public constructors in public API. Especially it's important to explain the different empty columns vs no columns methods/constructors.

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: +1

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 11]
```diff
@@ -69,10 +71,40 @@ public RowColumns(String row, String... columns) {
   }
 
   @SuppressWarnings("unchecked")
-  public T add(String... columns) {
+  public T add(String firstColumn, String... moreColumns) {
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: as I already mentioned, this is really awkward if you have code that dynamically computes the set of columns.

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: forget this. That works with the previous method.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 12]
```diff
@@ -29,30 +31,30 @@
...
-    this.columns = new ArrayList<>();
   }
 
   public RowColumns(byte[] row, byte[]... columns) {
-    this(row);
-    add(columns);
+    this.row = row;
+    this.columns = Arrays.asList(columns);
+  }
+
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: why do we have String versions of add() but not of the constructor?

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Oh we do. They come after the add() methods. Please rearrange the code to have all constructors first

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 10 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: So we protect against empty columns in add() but not in the constructor? That is not consistent. It should be the same in both places

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 13]
```diff
@@ -127,24 +129,34 @@ public void startTx(Transaction tx) {
...
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      Result[] hbaseResults = hTable.get(hbaseGets);
+
+      List<Map<byte[], byte[]>> results = new ArrayList<>(gets.size());
+      int hbaseResultsIndex = 0;
+      for (co.cask.cdap.api.dataset.table.Get get : gets) {
+        List<byte[]> cols = get.getColumns();
+        if (cols != null && cols.isEmpty()) {
+          results.add(ImmutableMap.<byte[], byte[]>of());
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: We could use a constant instead of creating a new ImmutableMap?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 10 above)
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: If user doesn't specify columns (a second parameter) in constructor, user will be using the constructor with only rowkey.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java (hunk 13 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: `ImmutableMap.of()` does use a constant, so no new instance is create.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 12 above)
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: moved.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 14]
```diff
@@ -29,50 +31,122 @@
...
-
-  public List<byte[]> getColumns() {
-    return columns;
-  }
+  private List<byte[]> columns;
 
   // key & column as byte[]
 
+  /**
+   * Defines a row, including all columns, unless additional columns are specified.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: The "unless" part is unclear. Maybe better be something like "Defines a row with all columns included. If any of the {@code add} method is called, only columns specified through those {@code add} method calls are included."

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 15]
```diff
@@ -29,50 +31,122 @@
...
-    Collections.addAll(this.columns, columns);
-    return (T) this;
+    this.row = row;
+    this.columns = Arrays.asList(columns);
   }
 
   // key and column as String
 
+  /**
+   * Defines a row, including all columns, unless additional columns are specified.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Same as above. Also good to specific the provided String row key is converted to `byte[]` with UTF-8 encoding.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 16]
```diff
@@ -29,50 +31,122 @@
...
+   * @param row the row to define.
+   */
   public RowColumns(String row) {
     this(Bytes.toBytes(row));
   }
 
+  /**
+   * Defines a row, including all columns, unless additional columns are specified.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Similarly, good to mention the provided strings are converted to `byte[]` with UTF-8 encoding.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 17]
```diff
@@ -29,50 +31,122 @@
...
+   * @param row the row to define
+   * @param columns the set of columns of the row to define
+   */
   public RowColumns(String row, String... columns) {
     this(row);
     add(columns);
   }
 
+  /**
+   * Specifies column(s) to be defined.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: What do mean by "to be defined"? Shouldn't it be "to be included"?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 18]
```diff
@@ -29,50 +31,122 @@
...
+  @SuppressWarnings("unchecked")
+  public T add(byte[][] columns) {
+    initArray();
+    Collections.addAll(this.columns, columns);
+    return (T) this;
+  }
+
+  /**
+   * Specifies column(s) to be defined.
+   * @param firstColumn the first column being defined
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Document the UTF-8 encoding conversion.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/Table.java [hunk 19]
```diff
@@ -203,7 +205,7 @@
    * multiple times (especially in transactions that delete many columns of the same rows).
    *
    * @param row row to delete from
-   * @param columns names of columns to delete
+   * @param columns names of columns to delete; if empty, will delete no columns
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/Table.java]: Better say "nothing will be deleted".

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java [hunk 20]
```diff
@@ -29,50 +31,127 @@
...
+
+  /**
+   * Specifies column(s) to be included.
+   * Note that the provided strings will be converted to {@code byte[]} with UTF-8 encoding.
+   * @param firstColumn the first column being included
+   * @param moreColumns a collection of additional columns being included
+   * @return the {@link RowColumns} object being defined
+   */
   @SuppressWarnings("unchecked")
-  public T add(String... columns) {
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: This is not backwards-compatible with the existing API. Is that a good idea in a bugfix release?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java [hunk 21]
```diff
@@ -496,9 +496,8 @@ private void putInternal(byte[] row, byte[][] columns, byte[][] values) {
   }
 
   /**
-   * NOTE: Depending on the use-case, calling this method may be much less
-   *       efficient than calling same method with columns as parameters because it may always require round trip to
-   *       persistent store
+   * NOTE: Depending on the use-case, calling this method may be much less efficient than calling same method
+   *       with columns as parameters because it may always require round trip to persistent store.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java]: may require a round trip? why is that?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 20 above)
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: I added a method below it to allow backwards-compatibility.
Does that resolve any code sample that would no longer work? Or what code sample would no longer work (is this breaking)?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 22]
```diff
@@ -127,24 +129,34 @@ public void startTx(Transaction tx) {
...
-        public Map<byte[], byte[]> apply(Result result) {
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      Result[] hbaseResults = hTable.get(hbaseGets);
+
+      List<Map<byte[], byte[]>> results = new ArrayList<>(gets.size());
+      int hbaseResultsIndex = 0;
+      for (co.cask.cdap.api.dataset.table.Get get : gets) {
+        List<byte[]> cols = get.getColumns();
+        if (cols != null && cols.isEmpty()) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: for code readability, I personally would have the same condition as in the the previous loop. I know this is the inverse condition, but whoever reads this code, needs to think twice to see that.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java [hunk 23]
```diff
@@ -296,7 +308,15 @@ private void setFilterIfNeeded(ScanBuilder scan, @Nullable Filter filter) {
     }
   }
 
+  /**
+   * Creates an {@link Get} for the specified row and columns.
+   *
+   * @param row the rowkey for the Get
+   * @param columns the columns to fetch. null means to retrieve all columns.
+   * @throws IllegalArgumentException if columns has lenght 0.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java]: typo: length.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java (hunk 21 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java]: This is a pre-existing comment, but I believe the reason is that the existing set of rows has to be read, and then a delete marker needs to be `PUT` to those columns (can't simply use hbase delete on the entire row).

Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java]: ok. I wonder whether that can be changed now. I believe that Tephra's TransactionalTable does use row deletes. Belongs into a different PR, though. However, you should remove the "may" because it is certain that it requires a round trip.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 20 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: existing APIs allowed calling add() without arguments. I don't think that is still possible.

**Comment by anew** [comment]
just a few final comments. The major one being about backwards-compatibility.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java (hunk 20 above)
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Would it make sense to have a method `public T add()`, marking it as deprecated in this PR. Then just that method can be removed.

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Even then it breaks compatibility with existing, compiled apps. Try it out. I think even for the case of 1+ arguments, the signature of the add() has changed and you will get a runtime error when executing the app.

Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Hmm, good point. Even though we do say users will have to recompile and redeploy after an upgrade, perhaps we should push this change to a non-patch version (i.e. to 3.4.0).

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: For compiled code, if you have a method that looks like:

`void add(byte[]...args);`

and then call it with `add()`, the compiler generate code like this:

```java
add(new byte[0]);
```

so, for binary compatibility, as long as you still have a `add(byte[][] bytes)`, it should work. For source compatibility, adding a `void add()` should solve the problem (should mark that as deprecate as well).

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: I prefer if you could test that before you merge. "should" is nice but better validate:)

Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Will test it.

Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: Actually, I was testing it out, and its not possible for user to call the no-parameter `add` method:

image

Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java]: I also tested the following code and it worked with the API changes.
```
cols = new byte[][] { ... };
get.add(cols);
```

**Comment by anew** [comment]
LGTM if tested.

**Comment by anwar6953** [comment]
Removed the last commit:
a97d9431c1b2e42109cd0adf332f1439be4351d8

**Comment by anew** [comment]
LGTM

**Comment by anwar6953** [comment]
Will run a final integration test against this code, and then merge.
